syntax = "proto3";

package messenger;

service MessengerServer {
    rpc MessageStream (Key) returns (stream Text);
    rpc SendMessage (stream Text) returns (stream SendConfirmation);
    rpc ValidateUser (User) returns (SendConfirmation);
    rpc RegisterKey (Key) returns (Empty);
    rpc ReturnUsers (User) returns (User);
}

message User {
    string user = 1;
}

message Text {
    string sender = 1;
    string receiver = 2;
    bytes cipher_text = 3;
    bytes tag = 4;
    bytes cipher = 5;
}

message Key {
    string key = 1;
}

message Empty {}

message SendConfirmation {
    string message = 1;
    string user_list = 2;
}